src/aiida/cmdline/commands/cmd_process.py

```diff
@@ -111,7 +120,9 @@ def process_list(
     if group:
         relationships['with_node'] = group
 
-    builder = CalculationQueryBuilder()
+    builder = CalculationQueryBuilder(
+        mapper=CalculationProjectionMapper(CalculationQueryBuilder._valid_projections, coloring=color)
+    )
```

Comment:
Still, the current solution is also quite weird and requires passing the `coloring` argument through to many steps. I see that it is tricky because it somehow needs to end up in the `format_state` function. Ideally the formatting is left to the actual command that prints the output, but that would have to be in `aiida.cmdline.commands.cmd_process` and that would also be quite ugly I am afraid.

Another thing I don't really like is that now `aiida.tools` imports `click` which is a command line utility. Normally, that should be kept to `aiida.cmdline`. But since it is just using the `style` method, it is not that bad. I wonder what these ansi codes render like in something that is not a terminal, e.g., a Jupyter notebook. Does it render properly?

With that all in mind. I wonder if there is a better solution. What if we just have `format_state` _always_ add the color through ANSI escape codes. Although `click` [does not (yet) support the `NO_COLOR` environment variable](https://github.com/pallets/click/issues/2207) there is an (undocumented) property of the context `color` that can be set to `False` to disable all ANSI color codes. So what you could do in `cmd_process` is simply
```
def command(ctx, ..., color):
    ctx.color = color
```
And that's it. If the flag is turned off, _all_ coloring is removed. This would allow to get rid off all the ugly passing through of `color`.

It would also make it easy to implement support for the `NO_COLOR` env variable as was mentioned by @danielhollas . What do you think?

Reply:
> I think this one is independent of this question, since the effect of this variables is global so it makes sense to me to implement them in src/aiida/__main__.py as suggested in https://github.com/pallets/click/issues/2207#issuecomment-1053562285
I just implemented it like this

Except I don't think this will always work. It will work when you invoke the CLI through `python -m aiida` which will route through `src/aiida/__main__.py`. However, it won't work when you call `verdi` directly. That is exposed as an entrypoint and directly calls `src/aiida/cmdline/commands/cmd_verdi.py::verdi`.

> Do I understand you correctly that we temporary disable it globally to get the formatting without colors, so we set it back to the original value afterwards so the echo_reports are printed with color?

It is a good question whether we even want to make a distinction here. If we provide a color option to a command, should it apply to _all_ output? I actually think that makes sense. So instead of adding a one-off option to `verdi process list` here, maybe we should think to have `--color` be a global option that applies to all commands (just like `--verbosity`). When specified, it will switch of all coloring. To me this doesn't only make more sense, but it will also be easier and cleaner to implement. Instead of having to pass through `color` arguments throughout the implementation, it can now be handled in one place and just configures the `click.Context` so it can automatically handle the formatting and strip color codes.

Reply:
Thanks for the clarification! I think implementing it like you suggest it as flag that is available for every command makes to me the most sense. Some commands are slightly problematic like `verdi run` that cannot enforce the coloring to its subprocess (or at least it needs extra logic). But I think it is acceptable in such cases (I think `verdi restapi` would be another) that the `--no-color` option does not work properly for now. Almost all commands print something to the terminal that is colorized by the echo utils. Maybe I can adapt the help page for these specific commands (`verdi run` and `verdi restapi`), so it is clear that it does not work properly.